pinot/generative/torch_gvae/train.py

```diff
@@ -0,0 +1,119 @@
+        epoch_test_ap_score = 0.
+
+        for molecule in graph_data:
+            ################ DATA PREPARATION #########################
+            # get the adjacency matrix for the molecular graph
+            # Because torch.SparseTensor doesn't interoperate too well with numpy
+            # or scipy sparse matrix, we would need to work with scipy sparse matrix
+            # and convert to torch.tensor where needed
+            (adj_orig,
+                adj_norm,
+                adj_label,
+                adj_train,
+                train_edges,
+                val_edges,
+                val_edges_false,
+                test_edges,
+                test_edges_false,
+                node_features,
+                norm,
+                pos_weight,
+                measurement) = molecule
+
+            ################ Optimization ######################
+
+            model.train()
+            optimizer.zero_grad()
+            n_nodes, num_features = node_features.shape
+            assert(num_features == feat_dim)
+
+            recovered, mu, logvar = model.encode_and_decode(node_features, adj_norm)
+
+            # Compute the (sub-sampled) negative ELBO loss
+            loss = negative_ELBO(preds=recovered, labels=adj_label,
+                                mu=mu, logvar=logvar, n_nodes=n_nodes,
+                                norm=norm, pos_weight=pos_weight)
+            loss.backward()
+            cur_loss = loss.item()
+            optimizer.step()
+
+            hidden_emb = mu.data.numpy() # embedding/latent representation of the nodes
+
+            # Compute the ROC score (with respect to link prediction task)
+            roc_val, ap_val = get_roc_score(hidden_emb, adj_orig, val_edges, val_edges_false)
+            roc_test, ap_test = get_roc_score(hidden_emb, adj_orig, test_edges, test_edges_false)
+
+            # Update average training/validation ROC/ validation AP score in this epoch
+            total_epoch_tr_loss += 1./N_molecules * loss
+            epoch_val_ap_score += 1./N_molecules * ap_val
+            epoch_val_roc_score += 1./N_molecules* roc_val
+            epoch_test_roc_score += 1./N_molecules * roc_test
+            epoch_test_ap_score += 1./N_molecules * ap_test
+
+
+        print("Epoch:", '%04d' % (epoch + 1), "train_loss=", "{:.5f}".format(total_epoch_tr_loss),
+            "val_ap=", "{:.5f}".format(epoch_val_ap_score),
+            "val_roc={:.5f}".format(epoch_val_roc_score),
+            "test_ap=", "{:.5f}".format(epoch_test_ap_score),
+            "test_roc={:.5f}".format(epoch_test_roc_score),
+            "time=", "{:.5f}".format(time.time() - t), 
+        )
+
+    print("Optimization Finished!")
+
+    # roc_score, ap_score = get_roc_score(hidden_emb, adj_orig, test_edges, test_edges_false)
+    # print('Test ROC score: ' + str(roc_score))
+    # print('Test AP score: ' + str(ap_score))
+
+
+if __name__ == '__main__':
+    gae_for(args)
```

Comment:
This training script looks great! But maybe move it to somewhere in `/scripts`? I think @karalets would argue that we need to find ways to organically integrate these experiments.

Reply:
I have a gvae_exp.py in `/scripts`, will remove this one since both are actually the same file. Thought of leaving `train.py` so that the module can be sort of a standalone module